### PR TITLE
os, build: Add support for loadable XIP_ELF offset calculation

### PR DESCRIPTION
**Pull request by edwakuwaku**
- mkldscript for loadable is generic tool, to break the dependency, chip specific should implement the logic independently
- If trpk file contains multiple binary, the logic should be applied
- Add flash virtual starting address and trpk multiple binary macros in defconfig

### CODE REVIEW

**Comment by edwakuwaku** [comment]
Hi @abhishek-samsung , could you please help to verify this PR on internal branch?

File: build/tools/amebasmart/gnu_utility/loadable_xip_elf.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+filepath = os.path.dirname(__file__) + "/km0_km4_app.bin"
+alignment = 4096
+
+def get_offset():
+    bin_size = os.path.getsize(filepath)
+    # Binary need to be aligned to 4K
+    bin_size = (bin_size + alignment - 1) // alignment * alignment
+    # 4K padding before AP binary (ie. running on Tizen Lite) starts
+    bin_size = bin_size + alignment
+    return hex(bin_size)
```
Comment (abhishek-samsung) [inline build/tools/amebasmart/gnu_utility/loadable_xip_elf.py:30]: I have a minor question. Does this also handle the size of Samsung header that we add at the start of the trpk file? If it doesn't then we will have issues at some cases where the alignment goes wrong.

**Review by abhishek-samsung — COMMENTED** [review]
@edwakuwaku , this PR works fine on the internal branch.

File: build/tools/amebasmart/gnu_utility/loadable_xip_elf.py (hunk 1 above)
Comment (edwakuwaku) [reply build/tools/amebasmart/gnu_utility/loadable_xip_elf.py:30]: Yes, the km0_km4 binary has a 4kb of header which is same as trpk file, despite of the values. Which means calculating from km0_km4 will yield the correct result.
In detail: 4KB header -> 4KB km4_km0 manifest -> km4_km0 image size with 4KB alignment -> 4KB ca32 manifset
Thus, your concern should be handled in the above logic.

Comment (abhishek-samsung) [reply build/tools/amebasmart/gnu_utility/loadable_xip_elf.py:30]: Okay, thanks for explaining

Comment (edwakuwaku) [reply build/tools/amebasmart/gnu_utility/loadable_xip_elf.py:30]: If there is no further question, could you leave an approval? Thanks~

File: build/tools/amebasmart/gnu_utility/rtk_data_binary.py [hunk 2]
```diff
@@ -1,4 +1,21 @@
 #!/usr/bin/env python
+###########################################################################
```
Comment (abhishek-samsung) [inline build/tools/amebasmart/gnu_utility/rtk_data_binary.py:2]: minor comment, this change doesn't seem to fit with the commit. Could it be moved to a new commit?

**Review by abhishek-samsung — APPROVED** [review]

File: build/tools/amebasmart/gnu_utility/rtk_data_binary.py (hunk 2 above)
Comment (edwakuwaku) [reply build/tools/amebasmart/gnu_utility/rtk_data_binary.py:2]: done~

**Review by abhishek-samsung — APPROVED** [review]

File: os/board/rtl8730e/Kconfig [hunk 3]
```diff
@@ -20,6 +20,17 @@ config FLASH_START_ADDR
...
+config TRPK_CONTAINS_MULTIPLE_BINARY
+    bool
+    default n
+    depends on XIP_ELF
+    help
+        This macro is to differentiate whether the kernel trpk contains
+        more than 1 binary. If yes, we have to recalculate the offset
+        according to the binary that is running on Tizen Lite, to make
+        sure the loadable_apps for XIP_ELF is placed to the correct
+        virtual flash address.
```
Comment (sunghan-chang) [inline os/board/rtl8730e/Kconfig:32]: The description is the same with the flat. But we don't need this for the flat. Are you able to leave why `XIP_ELF` requires this more?

File: build/configs/rtl8730e/loadable_apps/defconfig [hunk 4]
```diff
@@ -313,6 +313,8 @@ CONFIG_FLASH_MINOR=0
 CONFIG_FLASH_PART_SIZE="60,40,12,400,1844,4808,384,1844,4808,384,512,1280,8,"
 CONFIG_FLASH_PART_TYPE="none,none,none,none,kernel,bin,bin,kernel,bin,bin,smartfs,ftl,bootparam,"
 CONFIG_FLASH_PART_NAME="bl1,reserved,ftl,ss,kernel,app1,app2,kernel,app1,app2,userfs,reserved,bootparam,"
+CONFIG_TRPK_CONTAINS_MULTIPLE_BINARY=y
+CONFIG_FLASH_VSTART_LOADABLE=0xe000000
```
Comment (sunghan-chang) [inline build/configs/rtl8730e/loadable_apps/defconfig:317]: I think this defconfig does not enable `XIP_ELF` which you make a dependency. 
@abhishek-samsung Could you guide him which defconfig should be enabled?

Comment (edwakuwaku) [reply build/configs/rtl8730e/loadable_apps/defconfig:317]: But in https://github.com/Samsung/TizenRT/pull/6185, the loadable configs will apply XIP_ELF.
Also as I have known, there seems no more loadable ram cases for RTL8730E in the future, due to ram shortage.

Comment (abhishek-samsung) [reply build/configs/rtl8730e/loadable_apps/defconfig:317]: > @abhishek-samsung Could you guide him which defconfig should be enabled?

in the current code loadable_apps_xip/defconfig is the only one using xip apps. So changing only that particular config should be enough. I think #6185 can apply these changes required in that PR self.

Comment (edwakuwaku) [reply build/configs/rtl8730e/loadable_apps/defconfig:317]: okay, then I will only update to loadable_app_xip/defconfig

File: os/board/rtl8730e/Kconfig (hunk 3 above)
Comment (edwakuwaku) [reply os/board/rtl8730e/Kconfig]: Hi @sunghan-chang , could you check the comments again?

File: os/board/rtl8730e/Kconfig [hunk 5]
```diff
@@ -20,6 +20,21 @@ config FLASH_START_ADDR
         A start address of flash (in hex).
         This is fixed value, so user doesn't need to change it.
 
+config TRPK_CONTAINS_MULTIPLE_BINARY
+    bool
+    default n
+    depends on XIP_ELF
+    help
+        This macro is to differentiate whether the kernel trpk contains
```
[the remark this answers is not in the document]
Comment (edwakuwaku) [reply os/board/rtl8730e/Kconfig:28]: Yes, but is there any conflict in between? We used kernel trpk to calculate the offset for common and app

[the remark this answers is not in the document]
Comment (sunghan-chang) [reply os/board/rtl8730e/Kconfig:28]: Oh ok. Then the common and the app get affected from the kernel that includes km4. Right?

[the remark this answers is not in the document]
Comment (edwakuwaku) [reply os/board/rtl8730e/Kconfig:28]: yes correct, but we won't state "km4/km0" here, that's why this is just a bool to differentiate whether kernel trpk contains multiple binary

**Review by sunghan-chang — APPROVED** [review]

**Pull request merged**